ydb/library/actors/wilson/wilson_uploader.cpp

```diff
@@ -267,29 +279,43 @@ namespace NWilson {
                 SpansSizeBytes -= batch.SizeBytes;
 
                 ScheduleWakeup(NextSendTimestamp);
-                Context = std::make_unique<grpc::ClientContext>();
+
+                auto context = std::make_unique<grpc::ClientContext>();
                 if (GrpcSigner) {
-                    GrpcSigner->SignClientContext(*Context);
+                    GrpcSigner->SignClientContext(*context);
                 }
-                Reader = Stub->AsyncExport(Context.get(), std::move(batch.Request), &CQ);
-                Reader->Finish(&Response, &Status, nullptr);
+                auto reader = Stub->AsyncExport(context.get(), std::move(batch.Request), &CQ);
+                auto uploadData = new TExportRequestData {
+                    .Context = std::move(context),
+                    .Reader = std::move(reader),
+                };
+                uploadData->Reader->Finish(&uploadData->Response, &uploadData->Status, uploadData);
+                ALOG_TRACE(WILSON_SERVICE_ID, "started export request " << (void*)uploadData);
+                ExportRequests.PushBack(uploadData);
+                ++ExportRequestsCount;
             }
 
-            void CheckIfDone() {
-                if (Context) {
-                    void *tag;
-                    bool ok;
-                    if (CQ.AsyncNext(&tag, &ok, std::chrono::system_clock::now()) == grpc::CompletionQueue::GOT_EVENT) {
-                        if (!Status.ok()) {
-                            ALOG_ERROR(WILSON_SERVICE_ID,
-                                "failed to commit traces: " << Status.error_message());
-                        }
-
-                        Reader.reset();
-                        Context.reset();
-                    } else {
-                        ScheduleWakeup(TDuration::MilliSeconds(100));
+            void ReapCompletedRequests() {
+                if (ExportRequests.Empty()) {
+                    return;
+                }
+                void* tag;
+                bool ok;
+                while (CQ.AsyncNext(&tag, &ok, std::chrono::system_clock::now()) == grpc::CompletionQueue::GOT_EVENT) {
+                    auto node = static_cast<TExportRequestData*>(tag);
+                    ALOG_TRACE(WILSON_SERVICE_ID, "finished export request " << (void*)node);
+                    if (!node->Status.ok()) {
+                        ALOG_ERROR(WILSON_SERVICE_ID,
+                            "failed to commit traces: " << node->Status.error_message());
                     }
+                    
+                    --ExportRequestsCount;
+                    node->Unlink();
+                    delete node;
+                }
+
+                if (!ExportRequests.Empty()) {
+                    ScheduleWakeup(TDuration::MilliSeconds(100));
```

Comment:
by the way, how will this work with a big rate of spans?
may be we need to run another thread polling CQ and sending messages to the uploader?